### PR TITLE
Investigating Spectral Bias on Functional Autoencoder

### PR DESCRIPTION
**Pull request by stevengogogo**
## Objective

This work investigate the functional autoencoder that encode-decode functional data. This is a neural approach for functional principle  component analysis. The data is randomly generated from function that is infinite-diemensional domain. I suspect the functional autoencoder has same issue as regression problem that has spectral bias issue. 

## Method

The functional autoencoder uses basis fucntions decomposition to encoder function:

Suppose we have a function $u: \mathbb{R}\to \mathbb{R}$, and we want to decompose $u$ into 

$$
u \approx \sum_{i}^{n} c_i \phi_n(x)
$$

where the basis function $\phi_i$ for $i=1,\cdots, n$ is given by a neural network $NN: \mathbb{R} \times \Omega->\mathbb{R}$

$$
\phi_i(x) = NN_i(x; \theta)
$$

The score of coefficient $c_i$ is given by

$$
c_i = \langle \phi_i, u \rangle  = \int \phi_i(x) u(x) dx
$$

the integration is approximated by trapezoidal rule. The parameter $\theta$ is optimized by 

$$
loss = MSE(u, \hat{u}
$$

Also , the orthonormal property is induced by loss function.

## Related work

**Functional encoding on operator learning**
The functional encoding is a process to use finite dimensional domain to describe infinite dimensional domain. The Basis Operator Network uses functional encoder to derive the latent vector of target/input function. The advantage of this encoding method is sensor point $x$ can be in arbitrary resolution (not good for DeepONet) and irregular spacing (not good for FNO) once trained. As long as, the inner product is successfully estimated. The complexity of encoding is $O(n)$, and high dimensional space integration can be approximated by Monte Carlo estimation. 

**Multiscale**

The MultiscaleDNN can be applied to help mitigate the spectral bias. There are two ideas of doing this:

**Sequential layer**:  The autoencoder is in $l$ layer (or sequentially use same encoder),  first autoencoder get $\hat{u}_i$, and second autoencoder recover $\hat{u}_2 = u-\hat{u}_1$. The multilevel structure can also use scaling to decrease frequency spectrum or customized activation function for frequency modulation. The drawback of this structure is each layer need to apply sequential.
 
**Horizontal layer**: Use scaling series $[x, ..., nx]$ as input, and output multiple basis functions. The tricky part is how to make basis function orthonormal.

Reference:
1. "Basis operator network: A neural network-based model for learning nonlinear operators via neural basis"
2. "Deep Learning for Functional Data Analysis with Adaptive Basis Layers" ([arXiv](https://arxiv.org/abs/2106.10414))

## ToDo

To complete the implementation 

- [ ] Find a good problem to check the training works
- [ ] Apply orthonormal loss

To research

- [ ] add multiscale training or structure

### CODE REVIEW

**Comment by stevengogogo** [comment]
Copy post on Teams:

The[ preliminary resul](https://github.com/liruipeng/ml2/pull/14)t shows that the functional encoder has spectral bias. This might not be surprising since the it's still a DNN approach. I believe we can play some tricks like in MultiDNN or multi-step training on this. For example

Cascading functional encoder, and each AE learn to recover the residual function that previous AE produced. This can be same structure (like Residual NN) or different.
Scaling $[x,...,nx]$ like in multiDNN
Multi-step training the cascading functional encoder.

The ultimate goal is to produce fix dimensional latent factor of function to do operator learning with DeepONet.

**Comment by stevengogogo** [comment]
- Play with simple function

**Comment by stevengogogo** [comment]
## Exp: Simple function

The training can stay in same loss for a long time, and decay for a while

image
image

**Comment by stevengogogo** [comment]
## Exp: Train Functional Autoencoder up to a fixed freuqncy



image
image
image
image

**Comment by stevengogogo** [comment]
image
image
image

**Comment by stevengogogo** [comment]
## Exp: Traing with high frequency range

Functional autoencoder fails to train with high frequency=16

image
image

**Comment by stevengogogo** [comment]
## Exp: Multi-level training on multiscale problem

```
python pinn_1d.py   --levels 3 --epochs 20000 --lr 1e-3 --activation tanh --sweeps 1 --plot --hidden_dims 128 128 128 128   --loss_type 0 --high_freq 16 --loss_type 0 --bc_weight 500 --nx 2000  --gamma 1
```
```
Iteration    999/ 20000, PINN Loss: 3.3114e+06, Err 2-norm:  2.8072e+02, inf-norm: 2.1248e+01
Iteration   1999/ 20000, PINN Loss: 3.3070e+06, Err 2-norm:  2.7741e+02, inf-norm: 2.1024e+01
Iteration   2999/ 20000, PINN Loss: 3.3053e+06, Err 2-norm:  2.7713e+02, inf-norm: 2.1006e+01
Iteration   3999/ 20000, PINN Loss: 3.3061e+06, Err 2-norm:  2.7704e+02, inf-norm: 2.1000e+01
Iteration   4999/ 20000, PINN Loss: 3.3042e+06, Err 2-norm:  2.7708e+02, inf-norm: 2.1002e+01
Iteration   5999/ 20000, PINN Loss: 3.3058e+06, Err 2-norm:  2.7698e+02, inf-norm: 2.0993e+01
Iteration   6999/ 20000, PINN Loss: 3.3038e+06, Err 2-norm:  2.7707e+02, inf-norm: 2.1001e+01
Iteration   7999/ 20000, PINN Loss: 3.3039e+06, Err 2-norm:  2.7701e+02, inf-norm: 2.0998e+01
Iteration   8999/ 20000, PINN Loss: 3.3037e+06, Err 2-norm:  2.7703e+02, inf-norm: 2.0998e+01
Iteration   9999/ 20000, PINN Loss: 3.3033e+06, Err 2-norm:  2.7705e+02, inf-norm: 2.1000e+01
Iteration  10999/ 20000, PINN Loss: 3.3062e+06, Err 2-norm:  2.7713e+02, inf-norm: 2.0999e+01
Iteration  11999/ 20000, PINN Loss: 3.3037e+06, Err 2-norm:  2.7705e+02, inf-norm: 2.0998e+01
Iteration  12999/ 20000, PINN Loss: 3.3034e+06, Err 2-norm:  2.7705e+02, inf-norm: 2.1000e+01
Iteration  13999/ 20000, PINN Loss: 3.3043e+06, Err 2-norm:  2.7704e+02, inf-norm: 2.1000e+01
Iteration  14999/ 20000, PINN Loss: 3.3039e+06, Err 2-norm:  2.7705e+02, inf-norm: 2.0999e+01
Iteration  15999/ 20000, PINN Loss: 3.3033e+06, Err 2-norm:  2.7704e+02, inf-norm: 2.1000e+01
Iteration  16999/ 20000, PINN Loss: 3.3029e+06, Err 2-norm:  2.7705e+02, inf-norm: 2.1000e+01
Iteration  17999/ 20000, PINN Loss: 3.3029e+06, Err 2-norm:  2.7705e+02, inf-norm: 2.1000e+01
Iteration  18999/ 20000, PINN Loss: 3.3029e+06, Err 2-norm:  2.7705e+02, inf-norm: 2.1000e+01
Iteration  19999/ 20000, PINN Loss: 3.3029e+06, Err 2-norm:  2.7704e+02, inf-norm: 2.1000e+01

Training Level 1
Level 0: FROZEN
Level 1: TRAIN
Level 2: OFF
Iteration    999/ 20000, PINN Loss: 1.7350e+06, Err 2-norm:  1.3734e+02, inf-norm: 1.0304e+01
Iteration   1999/ 20000, PINN Loss: 1.7321e+06, Err 2-norm:  1.1457e+02, inf-norm: 8.6925e+00
Iteration   2999/ 20000, PINN Loss: 1.7310e+06, Err 2-norm:  1.0671e+02, inf-norm: 8.0850e+00
Iteration   3999/ 20000, PINN Loss: 1.7304e+06, Err 2-norm:  1.0435e+02, inf-norm: 7.9010e+00
Iteration   4999/ 20000, PINN Loss: 1.7325e+06, Err 2-norm:  1.0300e+02, inf-norm: 7.7851e+00
Iteration   5999/ 20000, PINN Loss: 1.7301e+06, Err 2-norm:  1.0125e+02, inf-norm: 7.6631e+00
Iteration   6999/ 20000, PINN Loss: 1.7305e+06, Err 2-norm:  1.0081e+02, inf-norm: 7.6222e+00
Iteration   7999/ 20000, PINN Loss: 1.7298e+06, Err 2-norm:  1.0062e+02, inf-norm: 7.6111e+00
Iteration   8999/ 20000, PINN Loss: 1.7299e+06, Err 2-norm:  1.0053e+02, inf-norm: 7.5995e+00
Iteration   9999/ 20000, PINN Loss: 1.7298e+06, Err 2-norm:  1.0039e+02, inf-norm: 7.5971e+00
Iteration  10999/ 20000, PINN Loss: 1.7298e+06, Err 2-norm:  1.0052e+02, inf-norm: 7.6014e+00
Iteration  11999/ 20000, PINN Loss: 1.7297e+06, Err 2-norm:  1.0048e+02, inf-norm: 7.6008e+00
Iteration  12999/ 20000, PINN Loss: 1.7299e+06, Err 2-norm:  1.0028e+02, inf-norm: 7.5963e+00
Iteration  13999/ 20000, PINN Loss: 1.7297e+06, Err 2-norm:  1.0050e+02, inf-norm: 7.6001e+00
Iteration  14999/ 20000, PINN Loss: 1.7296e+06, Err 2-norm:  1.0047e+02, inf-norm: 7.6020e+00
Iteration  15999/ 20000, PINN Loss: 1.7297e+06, Err 2-norm:  1.0048e+02, inf-norm: 7.5995e+00
Iteration  16999/ 20000, PINN Loss: 1.7296e+06, Err 2-norm:  1.0049e+02, inf-norm: 7.6032e+00
Iteration  17999/ 20000, PINN Loss: 1.7296e+06, Err 2-norm:  1.0047e+02, inf-norm: 7.6023e+00
Iteration  18999/ 20000, PINN Loss: 1.7296e+06, Err 2-norm:  1.0047e+02, inf-norm: 7.6026e+00
Iteration  19999/ 20000, PINN Loss: 1.7296e+06, Err 2-norm:  1.0052e+02, inf-norm: 7.6025e+00

Training Level 2
Level 0: FROZEN
Level 1: FROZEN
Level 2: TRAIN
Iteration    999/ 20000, PINN Loss: 5.5578e+05, Err 2-norm:  2.8064e+01, inf-norm: 2.5124e+00
Iteration   1999/ 20000, PINN Loss: 4.6835e+05, Err 2-norm:  1.6543e+01, inf-norm: 1.9554e+00
Iteration   2999/ 20000, PINN Loss: 4.5931e+05, Err 2-norm:  3.3680e+01, inf-norm: 3.5252e+00
Iteration   3999/ 20000, PINN Loss: 3.1723e+05, Err 2-norm:  2.0670e+01, inf-norm: 2.0088e+00
Iteration   4999/ 20000, PINN Loss: 1.1479e+05, Err 2-norm:  3.7247e+01, inf-norm: 3.4450e+00
Iteration   5999/ 20000, PINN Loss: 7.7330e+03, Err 2-norm:  3.5952e+01, inf-norm: 3.1789e+00
Iteration   6999/ 20000, PINN Loss: 2.7519e+02, Err 2-norm:  6.0936e+00, inf-norm: 5.4464e-01
Iteration   7999/ 20000, PINN Loss: 1.2830e+03, Err 2-norm:  1.1206e+00, inf-norm: 1.2224e-01
Iteration   8999/ 20000, PINN Loss: 6.4702e+01, Err 2-norm:  1.4265e-01, inf-norm: 1.4078e-02
Iteration   9999/ 20000, PINN Loss: 6.4180e+01, Err 2-norm:  2.6231e-02, inf-norm: 3.1756e-03
Iteration  10999/ 20000, PINN Loss: 4.0795e+02, Err 2-norm:  1.0429e+00, inf-norm: 1.0083e-01
Iteration  11999/ 20000, PINN Loss: 6.1485e+01, Err 2-norm:  6.1354e-03, inf-norm: 7.3051e-04
Iteration  12999/ 20000, PINN Loss: 6.0777e+01, Err 2-norm:  8.2555e-03, inf-norm: 8.8501e-04
Iteration  13999/ 20000, PINN Loss: 6.0145e+01, Err 2-norm:  7.1721e-03, inf-norm: 8.2356e-04
Iteration  14999/ 20000, PINN Loss: 5.9876e+01, Err 2-norm:  3.4553e-02, inf-norm: 3.6089e-03
Iteration  15999/ 20000, PINN Loss: 5.9284e+01, Err 2-norm:  3.9454e-03, inf-norm: 4.3374e-04
Iteration  16999/ 20000, PINN Loss: 5.8850e+01, Err 2-norm:  2.4648e-03, inf-norm: 2.6143e-04
Iteration  17999/ 20000, PINN Loss: 5.8467e+01, Err 2-norm:  2.5098e-03, inf-norm: 3.0148e-04
Iteration  18999/ 20000, PINN Loss: 5.8113e+01, Err 2-norm:  2.7034e-03, inf-norm: 2.8241e-04
Iteration  19999/ 20000, PINN Loss: 5.7784e+01, Err 2-norm:  3.2663e-03, inf-norm: 3.1281e-04
```
|Real|Frequency|
|---|---|
|image|image|
|image|image|
|image|image|

**Comment by stevengogogo** [comment]
## Exp: Fourier Feature embedding

The fourier feature embedding helps training better on training region. However, extrapolation remains bad.

The method is 

x->Fourier(x)->encoder-> Fourier(x)->Decoder

$$
B_{m\times D} \sim N(0, \sigma^2)
$$

I chose m=100

$$
\gamma(x) = [\cos(Bx), \sin(Bx)]^T
$$

image
image
image
image
image

**Comment by stevengogogo** [comment]
Functional autoencoder can not extrapolate different frequencies.
